storage/sql.go

```diff
@@ -0,0 +1,378 @@
+package storage
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/go-sql-driver/mysql"
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/telemetry"
+	tupleUtils "github.com/openfga/openfga/pkg/tuple"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	DefaultMaxTuplesPerWrite             = 100
+	DefaultMaxTypesPerAuthorizationModel = 100
+)
+
+type Datastore struct {
```

Comment:
It doesn't look like we benefit much from this abstraction and it adds a little structural indirection that I don't think is worth the abstraction. The shared code here is the underlying `*sql.DB`, `tracer`, `logger`, etc.. Everything but the underlying `*sql.DB` is shared database config, which we could simply provide a helper function for to bootstrap/initialize specific sql datastore implementations. My guiding rule behind this rationale is that this `Datastore` struct doesn't provide functionality in it's own right, and so it feels weird to introduce it just to share an abstraction over an existing abstraction (e.g. `*sql.DB`) and for config simplification which could easily be achieved with helper functions instead of a struct.

I like the changes that consolidate the iterators and the config options etc.. But I think we should move back towards specific datastore implementations here instead.

Reply:
Doesn't it allow us to have one entry point to construct a new Postgres or MySQL datastore? If we split it won't we need separate `WithTracer`, etc. functions as well? Or is that what you mean and that it is okay?

Reply:
Sorry, that wasn't clear. We would consolidate the storage options into `storage.go`but then leave the rest as it was. So essentially just de-duplicating the datastore options, but leaving the rest as is.